keps/sig-apps/3189-downscale-pod-picker/README.md

```diff
@@ -0,0 +1,858 @@
+
+### Risks and Mitigations
+
+- A slow `downscalePodPicker` API may increase the time to downscale a ReplicaSet
+   - MITIGATION: we will provide a `timeoutSeconds` config, if the API has not successfully returned within this time, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- The `downscalePodPicker` may not be reachable at all times, or may return an invalid response
+   - MITIGATION: we will provide a `maxRetries` config, if the API continues to fail after this many attempts, 
+     we proceed as if all Pods were "tied", and use the existing process to resolve the tie
+- Users may attempt to use `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` at the same time, resulting in unexpected behaviour
+   - MITIGATION: documentation will stress that only ONE of `downscalePodPicker` and `controller.kubernetes.io/pod-deletion-cost` should be used per ReplicaSet
+- A malicious (or poorly designed) `downscalePodPicker` API may attempt to overwhelm the controller by returning an extremely large payload
+   - MITIGATION: we will limit the maximum response size to something reasonable, like 1MB
+- Users may find it difficult to build their own `downscalePodPicker` API
+   - MITIGATION: documentation will provide example implementations in common programing languages
+- Users may create non-idempotent `downscalePodPicker` APIs that have side effects (e.g. preventing chosen workers from accepting new tasks):
+   - MITIGATION: documentation will stress that `downscalePodPicker` API must be idempotent, and may be called multiple times
+- Malicious actors may call `downscalePodPicker` APIs from outside the controller loop
+   - MITIGATION: documentation will stress that `downscalePodPicker` APIs should be secured by authentication at the network or request layer
+
+
+## Design Details
+
+#### K8S API Changes:
+
+A new `downscalePodPicker` field will be added to `ReplicaSetSpec` and `DeploymentSpec`,
+this field specifies a "Pod Picker" REST API that informs which Pods are removed when `replicas` is decreased.
+
+Here is an example ReplicaSet with thew new `downscalePodPicker` field:
+
+```yaml
+apiVersion: apps/v1
+kind: ReplicaSet
+metadata:
+  name: my-replicaset
+spec:
+  replicas: 3
+  selector:
+    matchLabels:
+      app: my-app
+  template:
+    metadata:
+      labels:
+        app: my-app
+    spec:
+      containers:
+        - name: my-container
+          image: my-image
+
+  #############################
+  ## this is the new section ##
+  #############################
+  downscalePodPicker:
+    http:
+      host: my-app.my-namespace.svc.cluster.local
+      httpHeaders:
+        - name: authentication
+          valueFrom:
+            secretKeyRef:
+              key: authentication-token
+              name: my-secret
+      path: "/downscale-pod-picker"
+      port: 443
+      scheme: "https"
```

Comment:
Why isn't this query made using the Kubernetes API? That would allow a simpler approach where, for example, the reference is to a custom resource that holds a static value, the name of the best victim Pod.

Because the Kubernetes API is extensible, the same design also allows a dynamic lookup.
An HTTP query to an external service, if needed, could be implemented as an APIService / extension API.

Reply:
@sftim what do you mean by "using the Kubernetes API"? This `downscalePodPicker.http.*` spec refers to an external service that will "pick" the pods which are best to kill.

If you are meaning why not have a field on some K8S resource called "podsToKill", this approach has all the same issues as the annotation approach:
- the information is not "on-demand", that is, we don't know when some external system will decrease the replicas, so who knows when to update this "podsToKill" field
- this will result in large numbers of PATCH calls if someone keeps it up to date in real-time (and probably DDOS the kube-apiserver)

Reply:
@alculquicondor @sftim If we were to add a `podsToDelete` field to the `spec` to `ReplicaSet`:

```yaml
apiVersion: apps/v1
kind: ReplicaSet
metadata:
  name: my-replicaset
spec:
  replicas: 3
  downscalePolicy:
    # how many seconds 
    threshholdSeconds: 120
    podsToDelete:
      - name: my-replicaset-aaaaaaaa-aaaaa
         lastUpdated: 1647409852
      - name: my-replicaset-aaaaaaaa-aaaaa
         lastUpdated: 1647409852
```

I see some significant problems:
- This violates the core K8S principle that the `spec` is a declaration of the user's desired "target" state. With this approach, `podsToDelete` is effectively a glorified message queue between the scaler and controller.
- `HorozontalPodAutoscaler` would need to be significantly updated with what is effectively a Pod-Picker API (otherwise users would be forced to create a custom autoscaler that updates `podsToDelete` before updating `replicas`)
- Pod names may be reused over time
- Unless the controller clears `podsToDelete` after every downscale, it will contain Pods that have already been deleted (the controller updating `spec` is a big no-no, only the user should update the `spec` of a resource)
- Whatever is updating `podsToDelete` would have to avoid adding Unassigned/PodPending/PodUnknown/Unready Pods (as they will always be killed by the controller first when downscaling)

----

In general, I think loose-coupling is the best way to design things, and a Pod-Picker API decouples the decision about what Pod is best to kill from all the other parts of Kubernetes.

Having the ReplicaSet controller "ask" the Pod-Picker API which Pods are best to kill, at the time when it wants to kill them,  cleanly separates the concerns of the controller (reconciling the real-world state with the current `spec`), from the concerns of the application (minimising disruption during a downscale).

Reply:
> If we were to add a `podsToDelete` field to the `spec` to ReplicaSet:

Surely you'd mean `status`? An instantaneous victim recommendation is closer to being an actual-state property than a desired-state property.

However, I don't like that at all and would like to boost the idea of a new extension API (and APIService, and sample code) to implement scale-down victim recommendations.

Reply:
> Then, Deployment controller could simply say "get the PodCost of all pods labelled foo=bar". It's fundamentally async, which I see as a net positive.

This APIService could implement PodCost as outlined in https://github.com/kubernetes/enhancements/pull/3190#pullrequestreview-868561920

Reply:
@sftim I assume you are suggesting we use `APIService` resources to remove the need for the ReplicaSet Controller to authenticate with the Pod-Picker API? 

I do quite like this idea, as it removes the need to bake API auth into the spec, and would allow us to simply have the ReplicaSet controller talk directly to [`/apis/myextension.mycompany.io/v1/…`](https://kubernetes.io/docs/concepts/extend-kubernetes/api-extension/apiserver-aggregation/#aggregation-layer) on the Kubernetes API.

But there are some big problems that stem from the `APIService` being a cluster-level thing that exposes an API under the Kubernetes API itself:
1. Users shouldn't need to create a new cluster-level `APIService` to use the Pod-Picker API on each namespaced ReplicaSet or Deployment
2. For large clusters with lots of Deployments, we would be significantly polluting the Kubernetes API path with lots of `APIServices`
3. Extension API Servers [have a complex approach](https://kubernetes.io/docs/tasks/extend-kubernetes/configure-aggregation-layer/#extension-apiserver-authenticates-the-request) to authenticate the `kube-apiserver` requests, which is tightly coupled to the Kubernetes cluster (making it more difficult for users to make their Pod-Picker API services)

Am I understanding your idea correctly?

----

Also, this discussion has made me realize we do need to add a [`caBundle`](https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.23/#webhookclientconfig-v1-admissionregistration-k8s-io) field to the `downscalePodPicker` spec, so that the `kube-apiserver` can verify the identity of the Pod-Picker API it's talking to.

Because with `httpHeaders`, we are only providing a method for the `kube-apiserver` to authenticate itself to the Pod-Picker API.

Reply:
If there are three or four different things that eventually get added to core Kubernetes, and are a bit like this proposed PodPicker, we'll have to work hard (or at least hard enough) to make the client configuration consistent, ensure that it works with NetworkPolicy, and other measures.

Building on APIService matches an existing pattern and I commend it _if_ we want an HTTP callout / lookup to be an optional part of workload scale-down.

It sounds like there is a challenge that APIService is cluster-level whereas the mooted PodPicker is namespace-level. However, that's also an interesting one - should access to a single namespace allow you to cause the control plane to make HTTP requests to arbitrary destinations? If that raises further questions, let's answer those.

Reply:
@sftim I am actually in Australia, so I probably won't be able to attend Kubecon, alternatively, we can organize an online whiteboard session in the next few weeks?

I have messaged you on the Kubernetes Slack, so we can try and organize something. 
(NOTE: timezones might be a bit painful)